### PR TITLE
[proxima-direct-io-inmem] fix seek to offset

### PR DESCRIPTION
**Pull request by je-ik**

### CODE REVIEW

File: direct/io-s3/src/main/java/cz/o2/proxima/direct/s3/S3Client.java [hunk 1]
```diff
@@ -53,6 +53,8 @@
 @Slf4j
 class S3Client implements Serializable {
 
+  private static final long serialVersionUID = 1L;
```
Comment (dmvk) [inline direct/io-s3/src/main/java/cz/o2/proxima/direct/s3/S3Client.java:56]: This doesn't belong to this PR.

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BeamCommitLogReader.java [hunk 2]
```diff
@@ -241,6 +243,10 @@ private BeamCommitLogReader(
         offset == null || !stopAtCurrent, "Offset can be used only for streaming reader");
   }
 
+  private URI getUri() {
```
Comment (dmvk) [inline beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BeamCommitLogReader.java:246]: unused?

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java [hunk 3]
```diff
@@ -63,6 +64,7 @@ private BlockingQueueLogObserver(String name, long limit, long startingWatermark
     this.watermark = new AtomicLong(startingWatermark);
     this.limit = limit;
     queue = new ArrayBlockingQueue<>(100);
+    log.debug("Created {}", this);
```
Comment (dmvk) [inline beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:67]: Can we have a better log message here?

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java [hunk 4]
```diff
@@ -75,24 +77,30 @@ public boolean onError(Throwable error) {
...
 
-  private boolean enqueue(StreamElement element, OnNextContext context) {
+  private boolean enqueue(StreamElement element, @Nullable OnNextContext context) {
     try {
       lastWrittenContext = context;
       if (limit-- > 0) {
         return putToQueue(element, context);
       }
-      log.debug("Terminating consumption of {} due to limit", name);
+      log.debug(
```
Comment (dmvk) [inline beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:102]: Maybe we should change this log message levels to trace?

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java [hunk 5]
```diff
@@ -132,7 +140,9 @@ private boolean putToQueue(@Nullable StreamElement element, @Nullable OnNextCont
 
   @Override
   public void onIdle(OnIdleContext context) {
-    updateAndLogWatermark(context.getWatermark());
+    if (queue.isEmpty()) {
```
Comment (dmvk) [inline beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:143]: 👍

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java [hunk 6]
```diff
@@ -168,6 +178,9 @@ StreamElement takeBlocking() throws InterruptedException {
   private StreamElement consumeTaken(@Nullable Pair<StreamElement, OnNextContext> taken) {
     if (taken != null && taken.getFirst() != null) {
       lastReadContext = taken.getSecond();
+      if (lastReadContext != null) {
```
Comment (dmvk) [inline beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:181]: did you actually get NPE here?

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java [hunk 7]
```diff
@@ -97,15 +98,21 @@
...
+    }
+
     @Getter final long watermark;
+    @Getter final Set<String> consumedKeyAttr;
 
-    public IntOffset(long offset, long watermark) {
-      this.offset = offset;
+    public ConsumedOffset(final Set<String> consumedKeyAttr, long watermark) {
+      synchronized (consumedKeyAttr) {
+        this.consumedKeyAttr = new HashSet<>(consumedKeyAttr);
```
Comment (dmvk) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:114]: java.util.Collections#unmodifiableSet

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java [hunk 8]
```diff
@@ -436,13 +455,15 @@ private void handleFlushDataBaseOnPosition(
             try {
               if (!killSwitch.get()) {
                 el = cloneAndUpdateAttribute(getEntityDescriptor(), el);
-                long off = offsetTracker.incrementAndGet();
                 watermark.update(el);
                 long w = watermark.getWatermark();
+                consumedOffsets.add(
+                    String.format("%s#%s:%d", el.getKey(), el.getAttribute(), el.getStamp()));
```
Comment (dmvk) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:461]: String format is expensive for hot code paths, can we replace this with simple string concat?

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java [hunk 9]
```diff
@@ -465,10 +486,12 @@ private void handleFlushDataBaseOnPosition(
               .sorted(Comparator.comparingLong(a -> a.getValue().getFirst()))
               .forEachOrdered(
                   e -> {
-                    if (restartedOffset.getAndIncrement() < offset.getOffset()) {
+                    String keyAttr = e.getKey().substring(prefixLength);
+                    long stamp = e.getValue().getFirst();
+                    if (offset.getConsumedKeyAttr().contains(keyAttr + ":" + stamp)) {
```
Comment (dmvk) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:491]: can you please describe under what circumstances this can happen in the inline comment?

File: direct/io-s3/src/main/java/cz/o2/proxima/direct/s3/S3Client.java (hunk 1 above)
Comment (je-ik) [reply direct/io-s3/src/main/java/cz/o2/proxima/direct/s3/S3Client.java:56]: It is not directly related to the commit, yes. I can split it into two commits, but that makes only a small difference. It is needed, because the write should have had the `serialVersionUID` defined. I'd keep it as it is.

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BeamCommitLogReader.java (hunk 2 above)
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BeamCommitLogReader.java:246]: no?

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java (hunk 3 above)
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:67]: What do you miss? The intent was to log that a new BlockingQueueLogObserver was created.

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java (hunk 6 above)
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:181]: yes

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java (hunk 8 above)
Comment (je-ik) [reply direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:461]: This is testing storage, performance is of no convern here.

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java (hunk 7 above)
Comment (je-ik) [reply direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:114]: Ah, actually no. That is not the same semantics. I need a clone.

File: beam/core/src/test/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserverTest.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+import cz.o2.proxima.direct.core.Partition;
+import cz.o2.proxima.repository.AttributeDescriptor;
+import cz.o2.proxima.repository.EntityDescriptor;
+import cz.o2.proxima.repository.Repository;
+import cz.o2.proxima.storage.StreamElement;
+import java.util.UUID;
+import org.junit.Test;
+
+/** Test {@link BlockingQueueLogObserver}. */
+public class BlockingQueueLogObserverTest {
```
Comment (dmvk) [inline beam/core/src/test/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserverTest.java:33]: 👍
